[GTK][WPE] Do not block while waiting for tiles in the threaded compositor

carlosgcampos opened the pull request:
#### aa1cd6738768d9d77157ff02722cedf683a8330b
<pre>
[GTK][WPE] Do not block while waiting for tiles in the threaded compositor
<a href="https://bugs.webkit.org/show_bug.cgi?id=303087">https://bugs.webkit.org/show_bug.cgi?id=303087</a>

Reviewed by Nikolas Zimmermann and Miguel Gomez.

Ensure we don&apos;t schedule a composition until all pending tiles are done
instead of synchronously waiting for tiles during the composition.
CoordinatedSceneState now keeps track of the ongoing tile painting
operations. A new state WaitingForTiles is added to the compositor so
that if there are pending tiles when a rendering update composition is
requested we switch to the WaitingForTiles state. The compositor is
notified by LayerTreeHost every time the pending tiles change, so if we
were waiting for tiles and there are no more pending tiles we switch to
Schedule state and continue normally. This is the previous behavior, but
asynchronously waiting for the tiles instead. This will allow to process
other composition requests while waiting for tiles of a rendering
update request.

Canonical link: <a href="https://commits.webkit.org/303564@main">https://commits.webkit.org/303564@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b6b4bab5c1ce9870926199bf81d4971cfc764189

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132851 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5352 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43938 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140384 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/84880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134721 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5704 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/5215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/101580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/84880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135797 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4003 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119023 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/82374 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83616 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/112887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/37141 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143038 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5021 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37726 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/109957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5103 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/5215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110134 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3839 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115289 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58541 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20592 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5075 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/33643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4913 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68527 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5165 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/5033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp:
```diff
@@ -166,19 +171,23 @@ Ref<SkiaRecordingResult> SkiaPaintingEngine::record(const GraphicsLayer& layer,
         auto* canvas = buffer->canvas();
         if (!canvas) {
             buffer->completePainting();
+            platformLayer->didPaintTile();
```
I think we could use a `ScopeExit` for this `platformLayer` (both here and in SkiaPaintingEngine.cpp)

carlosgcampos replied:
I'm not sure I understand the comment, this is already SkiaPaintingEngine.cpp, but I'm not sure a ScopeExit to call didPaintTile is worth it either. We could remove the early return to remove the duplicated calls, though

spenap replied:
The idea is to use a RAII object to avoid having to manually call `didPaintTile` at the end, and make it refactor-proof (so if the code layout changes or a new early return is added, it is called). 

There were two files or two methods where this platformLayer was used, sorry I wrote the wrong back link. It's up to you, Carlos. I'm also happy to submit a follow-up PR adding this myself

spenap commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedSceneState.h:
```diff
@@ -62,6 +62,10 @@ class CoordinatedSceneState final : public ThreadSafeRefCounted<CoordinatedScene
 
+    void willPaintTile();
+    void didPaintTile();
+    unsigned pendingTiles() { return m_pendingTiles.load(); }
```
This can be made const: https://en.cppreference.com/w/cpp/atomic/atomic/load. The atomic load is `const` from the pov of C++

nikolaszimmermann reviewed (approved):
Looks really nice - I'll approve it, you might want to use a RAII helper or ScopeExit to avoid forgetting to call didPaintTile, as @spenap suggested. I'd like @magomez to ack this as well.

carlosgcampos committed:
[GTK][WPE] Do not block while waiting for tiles in the threaded compositor
https://bugs.webkit.org/show_bug.cgi?id=303087

Reviewed by Nikolas Zimmermann and Miguel Gomez.

Ensure we don't schedule a composition until all pending tiles are done
instead of synchronously waiting for tiles during the composition.
CoordinatedSceneState now keeps track of the ongoing tile painting
operations. A new state WaitingForTiles is added to the compositor so
that if there are pending tiles when a rendering update composition is
requested we switch to the WaitingForTiles state. The compositor is
notified by LayerTreeHost every time the pending tiles change, so if we
were waiting for tiles and there are no more pending tiles we switch to
Schedule state and continue normally. This is the previous behavior, but
asynchronously waiting for the tiles instead. This will allow to process
other composition requests while waiting for tiles of a rendering
update request.

Canonical link: https://commits.webkit.org/303564@main
